tasks/task_manager.cc

```diff
@@ -117,48 +179,79 @@ bool task_manager::task::impl::is_done() const noexcept {
     return _status.state == tasks::task_manager::task_state::done;
 }
 
+future<std::vector<task_manager::task::task_essentials>> task_manager::task::impl::get_failed_children() const {
+    return _children.map_each_task<task_essentials>([] (const foreign_task_ptr&) { return std::nullopt; },
+        [] (const task_essentials& child) -> std::optional<task_essentials> {
+            if (child.task_status.state == task_state::failed || !child.failed_children.empty()) {
+                return child;
+            }
+            return std::nullopt;
+        }
+    );
+}
+
 void task_manager::task::impl::run_to_completion() {
-    (void)run().then_wrapped([this] (auto f) {
+    (void)run().then_wrapped([this] (auto f) -> future<> {
         if (f.failed()) {
-            finish_failed(f.get_exception());
+            co_await finish_failed(f.get_exception());
         } else {
+            std::exception_ptr ex = nullptr;
             try {
                 _as.check();
-                finish();
+                co_await finish();
             } catch (...) {
-                finish_failed(std::current_exception());
+                ex = std::current_exception();
+            }
+            if (ex) {
+                co_await finish_failed(ex);
             }
         }
     });
 }
 
-void task_manager::task::impl::finish() noexcept {
+future<> task_manager::task::impl::fold_into_parent() const noexcept {
```

Comment:
nit: since this is conditional, better call the function `maybe_fold_into_parent`